---
title: "9.5_clonotype_diversity_scRepertoire"
output: html_document
date: "2023-10-25"
---

scRepertoire is designed to take filter contig outputs from the 10x Genomics Cell Ranger pipeline, processes that data to assign clonotype based on two TCR or Ig chains, and analyze the clonotype dynamics. The latter can be separated into 1) clonotype-only analysis functions, such as unique clonotypes or clonal space quantification and 2) interaction with mRNA expression data using Seurat, SingleCellExperiment or Monocle 3 packages.

#Useful links
Useful link [scRepertoire vignette]: https://ncborcherding.github.io/vignettes/vignette.html 
scRepertoire comes with a data set derived from T cells from three patients with renal clear cell carcinoma to demonstrate the functionality of the R package. 

Useful link [Tidyverse cheat sheet]: https://www.datacamp.com/cheat-sheet/tidyverse-cheat-sheet-for-beginners Useful link [packCircles tutorial]: https://rdocumentation.org/packages/ggraph/versions/2.1.0/topics/pack_circles 


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#install.packages('xlsx') 
#install.packages('circlize') 
#install.packages('viridis') 
#install.packages('packcircles') 
#install.packages('ggplot2') 
```

#### Get the input file:
After the "4_Normalization_clustering_BasicViz" step, there should be a file named "data_clust_cc_stress_regressed.rds". Copy this file to the current folder.
Also,cpoy the'all_contig_annotations.csv' from CellRanger output to the currect folder


#### Load required packages:
To load the required packages, execute the following code:
```{r}
library(Seurat)
library(scRepertoire)
library(ggplot2)
library(readxl)
library(tibble)
library(dplyr)
library(patchwork)
library(ggrepel)
library(RColorBrewer)
library(tidyr)
library(ggpubr)
library(tidyverse)
library(scCustomize)
library(circlize) 
library(viridis) 
library(packcircles) 
library("xlsx")
library(svglite)
```

Note: If you haven't installed the packages yet, please install them before loading.

#### Step 1: Data Loading and pre-processing
```{r}
scRep_data <- readRDS("data_clust_cc_stress_regressed.rds")

scRep_data@meta.data
```

Now, Load the all contig file and then use createHTOContigList(). This function will return a list, separated by the group.by variable(s).
This function depends on the match of barcodes between the single-cell object and contigs. If there is a prefix or different suffix added to the barcode, this will result in no contigs recovered. As of right now, it is recommended you do this step before the integrated, as integration workflows commonly alter the barcodes.
```{r}
#Create contigs file from 'filtered_contig_annotations' 
contigs <- read.csv("all_contig_annotations.csv")

#Create contigs list combining contigs csv, Seurat obj, and group by 'hash.ID'
contig.list <- createHTOContigList(contigs, scRep_data, group.by = "hash.ID")

#Visualize contig_list data built into scRepertoire 
head(contig_list[[1]])

#Visualize imported contigs 
head(contigs)
```

As the output of CellRanger are quantifications of both the TCRA and TCRB chains, the next step is to create a single list object with the TCR genes (comprised of the VDJC genes) and CDR3 sequences by cell barcode. This is performed using the combineTCR(), where the input is the contig list. There is also the relabeling of the barcodes by sample and ID information to prevent duplicates.

The output of combineTCR() will be a list of contig data frames that will be reduced to the reads associated with a single-cell barcode. It will also combine the multiple reads into clonotype calls by either the nucleotide sequence (CTnt), amino acid sequence (CTaa), the VDJC gene sequence (CTgene) or the combination of the nucleotide and gene sequence (CTstrict). 
```{r}
#If you have biological replicates of two variables (ie B6 vs. SAP KO), make combined list of contig data that relabels barcodes according to sample and ID information to prevent duplicates. The 'sample' identifier will be your broad label whereas 'ID' will be your batch-specific or biological replicates.  

#Use 'removeNA' filter to remove any cell barcode with an NA value in at least one chain.

combined <- combineTCR(contig.list, samples = c("B6","B6","B6","B6","SAP","SAP","SAP","SAP"),
                        ID = c("B6-1", "B6-2", "B6-3", "B6-4", "SAP-1", "SAP-2", "SAP-3", "SAP-4"), removeNA = TRUE)

#Also can incorporate a stringent filter to remove any cells that have > 2 TCR chains 

# combined2 <- combineTCR(contig.list, samples = c("B6","B6","B6","B6","SAP","SAP","SAP","SAP"), ID = c("B6-1", "B6-2", "B6-3", "B6-4", "SAP-1", "SAP-2", "SAP-3","SAP-4"), removeNA = TRUE, removeMulti = TRUE

#Visualize 'combined' data -- should see CTnt, CTaa, CTgene, CTstrict columns added 
head(combined[[1]])

```

Some workflows will have the additional labeling of the standard barcode. Before proceeding, we can use the function stripBarcode() to avoid any labeling issues down the line. Importantly, stripBarcode() is for removing prefixes on barcodes that have resulted from other pipelines. See comment: 
```{r}
#No need for stripBarcode function if the barcodes look like: AAACGGGAGATGGCGT-1, AAACGGGAGATGGCGT.
#MUST PERFORM THIS FUNCTION AFTER COMBINETCR() WHEN ASSIGNING SAMPLE AND ID VARIABLES AS THIS CHANGES BARCODE FORMAT.

for (i in seq_along(combined)) {
    combined[[i]] <- stripBarcode(combined[[i]], column = 1, connector = "_", num_connects = 3)
}


#Visualize 'combined' data -- should see CTnt, CTaa, CTgene, CTstrict columns added 
head(combined[[1]])
```
Now, Embed combined TCR list to Seurat object. “strict” - uses the VDJC genes comprising the TCR/Ig + the nucleotide sequence of the CDR3 region. This is the proper definition of clonotype. Additionally, the clonotype call is trying to incorporate both loci, i.e, both TCRA and TCRB chains and if a single-cell barcode has multiple sequences identified (i.e., 2 TCRA chains expressed in one cell). Using the 10x approach, there is a subset of barcodes that only return one of the immune receptor chains, the unreturned chain is assigned an NA value.

In order to categorize the frequency, we have the variable proportion which if TRUE allows for the relative proportion or when FALSE will use absolute frequency to define clonotype groups.Frequency is calculated based on "group.by" variable. This can be left blank, which would result in frequencies calculated by individual run.  
```{r}
#Add 'combined' contig list to Seurat object. I am using clonocall "aa". It can also be set as "strict".

scRep_combined <- combineExpression(combined, 
                                     scRep_data, 
                                     cloneCall = "aa", 
                                     group.by = "hash.ID", 
                                     proportion = TRUE)
#View as metadata 
scRep_combined@meta.data

```
#### Calculate clonal diversity per cluster per hash.ID
Iterate through hash.ID to calculate clonal diversity per cluster from all the hash.IDs separately and make new sheets in the final appended .xlsx file.
The final appended .xlsx file can then be used to visualize different diversity indexes (per cluster) such as shannon, inverse sympson etc.
```{r}
Idents(scRep_combined) <- "hash.ID" # Set identifiers for 'hash.type'
levels(scRep_combined@active.ident) # Extract the levels.
hash_Ids <- levels(scRep_combined@active.ident) # Store hash.ID  as a variable for iteration.
for (j in seq_along(hash_Ids)) {
  Idents(scRep_combined) <- "hash.ID"
  scRep_subset <- subset(scRep_combined, idents = hash_Ids[j])
  
  # prep the combined seurat object to show cluster wise clonotype diversity
  combined_subset <- expression2List(scRep_subset, split.by = "seurat_clusters")
  
  ### save excel files. Here I am using clonecall "aa".
  diversity <- clonalDiversity(combined_subset, cloneCall = "aa", chain = "both", return.boots = FALSE, exportTable=TRUE)

write.xlsx(diversity, "clonalDiversity_B6vsSAP.xlsx", sheetName=hash_Ids[j], 
  col.names=TRUE, row.names=TRUE, append=TRUE)
}
```


 
Examine top clonotypes for each experimental group  
 
```{r}
#First subset by B6
Idents(scRep_combined) <- "hash.type"
B6_subset <- subset(scRep_combined, idents = "B6")



#First, group based on unique frequencies/unique clonotypes using subset data. Summarize function is used to remove duplicates and group based on total count. Arranged based on count in descending order to identify top clonotypes. 

data_groupB6 <- B6_subset@meta.data %>%
  group_by(seurat_clusters, CTgene, CTnt, CTaa, CTstrict, Frequency, cloneType) %>%
             summarise(count = n()) %>%
             arrange(desc(count))
#Remove any NA values by 'na.omit' function
data_groupB6 <- na.omit(data_groupB6)
#View(data_groupB6)

#Print top ten clonotypes based on amino acid sequence 
head(data_groupB6$CTaa, n=10)
```

```{r}
Idents(scRep_combined) <- "hash.type" 
SAP_subset <- subset(scRep_combined, idents = "SAP")

#First, group based on unique frequencies/unique clonotypes using subset data.Summarize function is used to remove duplicates and group based on total count. Arranged based on count in descending order to idetify top clonotypes

data_groupSAP <- SAP_subset@meta.data %>%
  group_by(.dots = c("seurat_clusters", "CTgene", "CTnt", "CTaa", "CTstrict", "Frequency", "cloneType")) %>%
             summarise(count = n()) %>%
             arrange(desc(count))
#Remove any NA values by 'na.omit' function
data_groupSAP <- na.omit(data_groupSAP)
#View(data_groupSAP)
#Print top ten clonotypes based on amino acid sequence 
head(data_groupSAP$CTaa, n=10)
```


###vVisualizing contigs 

Quantify contigs:vThe first function to explore the clonotypes is quantContig() to return the total or relative numbers of unique clonotypes.
```{r}
#TRUE - relative percent of unique clonotypes scaled by the total size of the clonotype repertoire. 

quantContig(combined, cloneCall="gene+nt", scale = T)
#Visualize by alpha-chain
TRG_contig <- quantContig(combined, cloneCall="gene+nt", chain = "TRG", scale = T)
#Visualize by beta-chain
TRD_contig <- quantContig(combined, cloneCall="gene+nt", chain = "TRD", scale = T)
#Visualize by 'both' - combined chain visualization
TCR_both_contig <- quantContig(combined, cloneCall="gene+nt", scale = T, chain = "both")

#Save figures 
ggsave(TRG_contig, path = "Figs", filename = "TRG_contig.png", height = 4, width =8, units = 'in', dpi = 300, bg = "transparent", device = 'png')
ggsave(path = "Figs", filename = "TRG_contig.svg", height = 4, width = 8, units = 'in', dpi = 300, bg = "transparent", device = 'svg')

ggsave(TRD_contig, path = "Figs", filename = "TRD_contig.png", height = 4, width = 8, units = 'in', dpi = 300, bg = "transparent", device = 'png')
ggsave(path = "Figs", filename = "TRD_contig.svg", height = 4, width = 8, units = 'in', dpi = 300, bg = "transparent", device = 'svg')

ggsave(TCR_both_contig, path = "Figs", filename = "TCR_both_contig.png", height = 4, width = 8, units = 'in', dpi = 300, bg = "transparent", device = 'png')
ggsave(path = "Figs", filename = "TCR_both_contig.svg", height = 4, width = 8, units = 'in', dpi = 300, bg = "transparent", device = 'svg')


```

Export unique clonotype as table: Within each of the general analysis functions, there is the ability to export the data frame used to create the visualization. To get the exported values, use exportTable == T.
```{r}

quantContig_output <- quantContig(combined, cloneCall="strict", 
                            scale = T, chain = "both")

head(quantContig_output[[1]])
```

#Compare clonotypes 
We can also look at clonotypes between samples and changes in dynamics by using the compareClonotypes() function.
Samples=hash.ID (as set earlier in R script). Format: "sample_ID" as identified prior.  
```{r}
#Numbers -- top number of clonotypes to graph, this will be calculated based on the frequency of the individual sample. This can also be left blank.
Compare <- compareClonotypes(combined, 
                  numbers = 5, 
                  samples = c("B6_B6-1","B6_B6-2","B6_B6-3","B6_B6-4","SAP_SAP-1","SAP_SAP-2","SAP_SAP-3","SAP_SAP-4"),
                  cloneCall="aa",
                  graph = "alluvial")

#Save figures 
ggsave(Compare, path = "Figs", filename = "compare_B6vsSAP.png", height = 4, width = 12, units = 'in', dpi = 300, bg = "transparent", device = 'png')
ggsave(path = "Figs",filename = "compare_B6vsSAP.svg", height = 4, width = 12, units = 'in', dpi = 300, bg = "transparent", device = 'svg')

```


##Overlap Analysis 
Measures of similarity between the samples loaded into scRepertoire.Three methods currently can be performed in clonalOverlap() 1) overlap coefficient, 2) Morisita index, or 3) Jaccard index. The former is looking at the overlap of clonotypes scaled to the length of unique clonotypes in the smaller sample. The Morisita index is more complex, it is an ecological measure of the dispersion of individuals within a population, incorporating the size of the population. The Jaccard Similarity Index is very similar to the overlap coefficient - instead of using the length of the smaller sample, the denominator for the Jaccard Index is the union of the two comparisons, leading to a much smaller number.
```{r}
# Split the object to prep for the comparison
combined_b6_sap <- expression2List(scRep_combined, split.by = "hash.ID")

o1 <- clonalOverlap(combined_b6_sap, 
              cloneCall = "aa",
              chain = "both",
              method = "overlap"
              )
ggsave(o1, path = "Figs", filename = "Overlap_clonaloverlap_B6vsSAP.png",  height=7, width=14, units='in', dpi = 300, bg = "transparent", device='png')
ggsave(o1, path = "Figs", filename = "Overlap_clonaloverlap_B6vsSAP.svg",  height=7, width=14, units='in', dpi = 300, bg = "transparent", device='svg')

e1 <- clonalOverlap(combined_b6_sap, 
              cloneCall = "aa",
              chain = "both",
              method = "overlap",
              exportTable = TRUE
              )

write.xlsx(e1, "clonaloverlap_B6vsSAP.xlsx", sheetName="overlap", 
  col.names=TRUE, row.names=TRUE, append=TRUE)
```


###Integration with single cell data 
```{r}
#Visualize previously clustered data 
DimPlot(scRep_data, group.by = "seurat_clusters", pt.size = 1.5, label = TRUE)

ggsave("seurat_clusters.png", height = 8, width = 12, units= "in", dpi = 300, bg= "transparent", device = "png")

#Clonotypic information has been attached to Seurat object using the combineExpression() --> 'scRep_combined' 

colorblind_vector <- colorRampPalette(rev(c("#0D0887FF", "#47039FFF", 
              "#7301A8FF", "#9C179EFF", "#BD3786FF", "#D8576BFF",
              "#ED7953FF","#FA9E3BFF", "#FDC926FF", "#F0F921FF")))

cloneType_DimPlot <- DimPlot(scRep_combined, group.by = "cloneType") +
    scale_color_manual(values = colorblind_vector(5), na.value="grey") + 
  theme(plot.title = element_blank())

#Save DimPlot 
ggsave(cloneType_DimPlot, path = "Figs", filename = "cloneType_DimPlot.png", height = 5, width = 10, units = 'in', dpi = 300, bg = "transparent", device = 'png')
ggsave(path = "Figs", filename = "cloneType_DimPlot.svg", height = 5, width = 10, units = 'in', dpi = 300, bg = "transparent", device = 'svg')

```
##clonalOverlay
Using the dimensional reduction graphs as a reference, we can also generate an overlay of the position of clonally expanded cells using clonalOverlay(). Select the reduction for the visualization, default is “PCA” and the freq.cutpoint or lowest clonal frequency or proportion to generate the contour plot. 

clonalOverlay() can be used to look across all cells or faceted by a metadata variable using facet. As we facet, the overall dimensional reduction will be maintained, while the contour plots will adjust based on the facet variable
```{r}
#Idents(scRep_combined) <- "hash.ID"
clonal_overlay <- clonalOverlay(scRep_combined, 
              reduction = "umap", 
              freq.cutpoint = 0.001, 
              bins = 10, 
              facet = "hash.ID") + 
                 guides(color = "none")

clonal_overlay2 <- clonalOverlay(scRep_combined, 
              reduction = "umap", 
              freq.cutpoint = 0.001, 
              bins = 10) + 
                 guides(color = "none")

#Save combined figure 
ggsave(clonal_overlay, path = "Figs", filename = "clonal_overlay.png", height = 10, width = 15, units = 'in', dpi = 300, bg = "transparent", device = 'png')
ggsave(path = "Figs", filename = "clonal_overlay.svg", height = 10, width = 15, units = 'in', dpi = 300, bg = "transparent", device = 'svg') 

ggsave(clonal_overlay2, path = "Figs", filename = "clonal_overlay2.png", height = 10, width = 15, units = 'in', dpi = 300, bg = "transparent", device = 'png')
ggsave(path = "Figs", filename = "clonal_overlay2.svg", height = 10, width = 15, units = 'in', dpi = 300, bg = "transparent", device = 'svg')

```



#Clonotype Bias 
A clonal metric that seeks to quantify how individual clones are skewed towards a specific cellular compartment or cluster.split.by: Variable used to calculate the baseline frequencies; group.by: The compartment/variable used for the purpose of the comparison. min.expand: clonotype frequency cut off for the purpose of comparison
```{r}
clonotypeBias(scRep_combined, 
              cloneCall = "strict", 
              split.by = "hash.ID", 
              group.by = "seurat_clusters",
              n.boots = 20, 
              min.expand =0.001)
```

